ext/ddtrace_profiling_native_extension/heap_recorder.c

```diff
@@ -177,12 +183,22 @@ void heap_recorder_free(heap_recorder *heap_recorder) {
     return;
   }
 
+  // Clean-up all object records
   st_foreach(heap_recorder->object_records, st_object_record_entry_free, 0);
   st_free_table(heap_recorder->object_records);
 
+  // Clean-up all heap records (this includes those only referred to by queued_samples)
   st_foreach(heap_recorder->heap_records, st_heap_record_entry_free, 0);
   st_free_table(heap_recorder->heap_records);
 
+  // Clean-up dangling object datas that are not part of object records yet
+  if (heap_recorder->active_recording.obj_id != 0) {
+    live_object_data_free(heap_recorder->active_recording.object_data);
+  }
+  for (size_t i = 0; i < heap_recorder->queued_samples_len; i++) {
+    live_object_data_free(heap_recorder->queued_samples[i].object_data);
+  }
```

Comment:
If we do keep the queue (as per discussion in earlier PRs), I wonder if we should consider having the `queued_samples` not be an array of structures, but instead an array of `object_record *`. We're going to do those allocations anyway, so we may as well do them as part of enqueuing, rather than waiting until later maybe?

Reply:
Killed the queue in part 3!